### PR TITLE
Improvements to managing database connections at login

### PR DESCRIPTION
**Pull request by isedwards**
The user can now select which database they want to connect to from a dropdown list that contains the names of all available database connections.

login

Users can click on "Manage database connections".  If they are not a Windows Administrator then they see the following message:

admin

Database connections are edited in a grid view. When the users clicks okay, the connection details are stored in the Windows Registry under `Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Climsoft4`. This ensures that the connection details are still available when the software is updated to a new version.

db_connections

### CODE REVIEW

**Comment by Patowhiz** [comment]
Can we add a checkbox for "remember password" on this form? For users who want the application to remember the password and username used in the last login. A user might want to uncheck if he is not using his/her own computer to login.

**Comment by mhabimana** [comment]
@Patowhiz - Thanks, Sounds like a great idea.

**Comment by isedwards** [comment]
I'm happy with an option to remember the most recent username, but I'm less sure about remembering the user's password as well.

@smachua what do you think?

**Comment by mhabimana** [comment]
@isedwards :I think remembering the recent username would be the best option because remembering the password can expose the system to an-authorises persons who can temper with the system.

**Comment by smachua** [comment]
I agree with @isedwards and @mhabimana  because in many cases I have seen one desktop being used by different Climsoft operators through the same Windows login at different times. It will then be tricky on which password to remember. However, remembering the recent user names will be useful.

**Comment by smachua** [comment]
@isedwards I am yet to test and find out if there will be a problem in using Windows Registry key; 'Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Climsoft4' by Climsoft administrators who are not Windows Administrators. I will report my findings.

**Comment by Patowhiz** [comment]
@iedwards why didn't you opt for the application data folder(AppData  - Roaming, Local, LocalLow)? The trend I see nowadays is more applications leaning towards it because of the many merits  and flexibility it offers. Each windows user has his/her own  application data folder(both administrators and standard users), and we can store the user settings in any size and any format that we want(since this is bound to happen as we design the settings options for CLIMSOFT).

**Comment by maxwellfundi** [comment]
We should not be remembering passwords I think because as @mhabimana says, it can expose the system to unauthorized users mostly in cases which @smachua mentions on multiple users loging in on same computer

**Comment by isedwards** [comment]
*This version resolves conflicts and also includes a "remember username" option.*

@Patowhiz - For the database connection settings, typically all users of a machine need to be provided with the same settings, so the registry key seem like a good choice. I've used the built-in `My.Settings` to store the remembered username, I think this is stored in AppData, and has the advantages that you mention, but the details are handled by .NET.
@smachua thank you for testing - I believe that the way it should work is that any user can read the registry key, but only Administrators can update the values.

**Comment by Iyakaremyevedaste** [comment]
This is a great idea. However to ensure the security of our databases, make it in a such way the system will only remember username not password.

**Comment by mhabimana** [comment]
@Iyakaremyevedaste :Thanks,this is noted.

**Comment by smachua** [comment]
@isedwards the problem comes when Climsoft administrator is not a Windows administrator on a given desktop. I have encountered few such cases in Uganda and Bhutan where desktops meant for other operations are sometimes allowed to be used for Climsoft.

**Comment by isedwards** [comment]
@smachua - do you have the same problem with the `config.inf` file?  The file can usually only be editted by a Winsows administrator because it is in the Program Files folder.  What would you do in that situation?

**Comment by smachua** [comment]
@isedwards  Windows allows the administrator to give write permission to all users who require to update that file (Climsoft users) even if they are not Windows administrators. We have described that procedure in the Climsoft manual. Am not sure if that can be done on a registry key.

**Comment by isedwards** [comment]
@smachua - yes, we can have a option in Climsoft (when setting up the software for the first time) that will allow any user to manage the database connections.  It will require one-off Windows Administrator access to set this option and then it will make the registry key writable by all users.

Does this sound okay?

To prevent users from accidentally changing the database settings, we could make it so that the Climsoft Administrator has to ctrl-click the `Manage database connections` link to access the settings when they are not a Windows Administrator... anyone who doesn't control-click will get the usual message that they have to be a Windows Administrator.

Ctrl-click will only work when the option to allow access by all users has been set.

**Comment by smachua** [comment]
@isedwards let me think about then I will respond.

File: ClimsoftVer4/ClimsoftVer4/frmLogin.vb [hunk 1]
```diff
@@ -20,206 +20,203 @@ Imports ClimsoftVer4.Translations
...
+            ' There are other locations in the software that may access this
+            txtusrpwd.Text = builder.ConnectionString & ";Convert Zero Datetime=True"
+        Catch ex As Exception
+            MsgBox("Login failed: " & ex.Message)
+        End Try
 
-            ' Contruct the connection string using the output the file
+        conn.ConnectionString = txtusrpwd.Text
+        Try
+            conn.Open()
```
Comment (Patowhiz) [inline ClimsoftVer4/ClimsoftVer4/frmLogin.vb]: @isedwards the connection is not being closed?

**Comment by smachua** [comment]
@isedwards the method you have suggested sounds ok though slightly complicated and may require documentation. I don't know what is the opinion of others but I am a bit uncomfortable with the use of registry in storing such commonly used settings. We used it in V3.2 but users found it a bit troublesome and recommended for a change. That is why we instead opted to using protected files in V4. For instance when uninstalling the software you have to work out ways of deleting the registry key because if another installation becomes necessary there will be a crash.

**Comment by mhabimana** [comment]
@isedwards: I agree with @smachua ,we should avoid storing settings in the registry and look for an alternative, otherwise we will end up with the same issues faced in previous version 3.2. ideally, the installation and configuration and uninstallation of Climsoft need to be easy,smooth and straightforward. 
If remember well,you can not install and configure Climsoft v3.2,if you are not the administrator of the PC or a member of the administrator. So, to overcome this,we used to upgrade the current user or create a new user profile with admin privilege to be able to install  and use Climsoft V3.2 and this use to raise many issues and complications to users because,you may find a PC with many users and each and every time you have to switch users.
To overcome such issues let see another place were the settings can be stored so that any PC user will be able to install,configure and use Climsoft without using the privilege of the PC admin.

**Comment by mhabimana** [comment]
> @smachua - do you have the same problem with the `config.inf` file? The file can usually only be editted by a Winsows administrator because it is in the Program Files folder. What would you do in that situation?

Good you mentioned this because this is also an other tricky part during the installation  of climsoft because, during the installation,you have to go to C:\Program Files (x86)\ClimsoftV4 and right click,security and give permission to the current user,otherwise,this user will not have access to resources in this path.

**Comment by Patowhiz** [comment]
I queried about the choice of the registry before the code was merged. And I explained a bit about the alternative. We have a recommended alternative that latest windows applications use to store applications settings. And the issue of rights is already addressed by microsoft in the alternative. 

You can get my explanation in this conversation thread.

I suggest in future we test and look at the PR implementation more critically before merging code not after merging code.

Again @iedwards  we don't have a github "group" which we can tag developers to review code changes once a PR or an issue is raised. At the moment I have to tag each one of you individually. I missed this PR when you first created it, I had to manually check on current PRs.

**Comment by smachua** [comment]
Merging of this PR may be reversed since its discussion was not exhausted.

**Comment by isedwards** [comment]
We have @opencdms/climsoft-developers and other groups set up (see [teams](https://github.com/orgs/opencdms/teams)).

The PRs are sometimes being merged early (before they are ready) as a way of testing that they work with the recent changes on master. This is bad practice - a much better way is to do a pull/merge locally - but this is more complicated.  If master has moved forward then all the PRs should be rebased before testing. Currently our [process for merging](https://github.com/opencdms/Climsoft/wiki/Open-Source-Contributions) only requires a reviewer to approve changes, so this probably needs updating/improving.

In this case - this PR did not need to be merged with master for testing because it already contained all the recent changes that are in master. I agree with @smachua that we should revert this until it is fully approved.

**Comment by isedwards** [comment]
Whether we use the registry or a file, ideally we're looking for a solution that makes it possible to:
- Install and use the Climsoft client on a computer without ever having Windows Administrator access
- Set up and manage database connection details without needing to be a Windows Administrator, but with some kind of protection to prevent accidental changes
- Make sure the connection details are still available after updating to the latest version of the software
- Not require users to have to follow any detailed instructions (e.g. editing folder permisions or registry permissions)

My understanding of AppData (Roaming, Local, LocalLow) is that these apply on a per user basis, and not to all users (e.g. a different user/group being able to write to the configuration file).

It may be possible for the software to use `C:\ProgramData` (aka `C:\Users\All Users`) to create a folder as the current user and, as the current user, to make this writable to all users?

**Comment by Patowhiz** [comment]
@isedwards correct. **AppData** is per **application basis** AND per **user basis**. 
Microsoft defines it as 

> App data is mutable data that is created and managed by a specific app. It includes runtime state, app settings, user preferences, reference content (such as the dictionary definitions in a dictionary app), and other settings.

You can read more about AppData [here](https://docs.microsoft.com/en-us/windows/uwp/design/app-settings/store-and-retrieve-app-data)

I recommended using it because I thought we want user specific settings to be stored in their specific microsoft windows account, then let windows synchronise the setting for us(It picks those settings in the **Roaming** folder and synchronises on all the devices that have the account).  Also it's important to note, the lifetime of the app data is tied to the lifetime of the app(Which I think is what we want?). 
As for the general settings - those done by the administrator, I thought they are to be stored in the server not in the client machines. And have them applicable to users during runtime. 

So in summary, I had thought, user specific settings goes to AppData (Maybe inside the Roaming folder? where the user will get those settings in any devices that has his/her account). Then the general settings get to be stored in a public folder in the server. The security layer being an encrypted file that can only be read by climsoft application, if these settings are sensitive.

If that's not the case then I think @isedwards suggestion is a better alternative compared to the registry(So Ideally, we will be using [User Data](https://docs.microsoft.com/en-us/windows/uwp/design/app-settings/store-and-retrieve-app-data) folder - which is used to store data that a user creates and manages when using an app).

File: ClimsoftVer4/ClimsoftVer4/frmLogin.vb (hunk 1 above)
Comment (isedwards) [reply ClimsoftVer4/ClimsoftVer4/frmLogin.vb]: Thank you Patrick - this is fixed in #575 ([here](https://github.com/opencdms/Climsoft/pull/575/files#diff-31e5ad9223abb094c0afcb16d231bcceR139))
